crates/re_space_view_time_series/src/line_visualizer_system.rs

```diff
@@ -0,0 +1,212 @@
+use crate::{overrides::lookup_override, PlotPoint, PlotPointAttrs, PlotSeries, PlotSeriesKind};
+
+/// The system for rendering [`SeriesLine`] archetypes.
+#[derive(Default, Debug)]
+pub struct SeriesLineSystem {
+    pub annotation_map: AnnotationMap,
+    pub lines: Vec<PlotSeries>,
+}
+
+impl IdentifiedViewSystem for SeriesLineSystem {
+    fn identifier() -> re_viewer_context::ViewSystemIdentifier {
+        "SeriesLine".into()
+    }
+}
+
+impl VisualizerSystem for SeriesLineSystem {
+    fn visualizer_query_info(&self) -> VisualizerQueryInfo {
+        let mut query_info = VisualizerQueryInfo::from_archetype::<archetypes::Scalar>();
+        let mut series_line_queried: ComponentNameSet = SeriesLine::all_components()
+            .iter()
+            .map(ToOwned::to_owned)
+            .collect::<ComponentNameSet>();
+        query_info.queried.append(&mut series_line_queried);
+        query_info
+    }
+
+    fn execute(
+        &mut self,
+        ctx: &ViewerContext<'_>,
+        query: &ViewQuery<'_>,
+        _context: &re_viewer_context::ViewContextCollection,
+    ) -> Result<Vec<re_renderer::QueueableDrawData>, SpaceViewSystemExecutionError> {
+        re_tracing::profile_function!();
+
+        self.annotation_map.load(
+            ctx,
+            &query.latest_at_query(),
+            query
+                .iter_visible_data_results(Self::identifier())
+                .map(|data| &data.entity_path),
+        );
+
+        match self.load_scalars(ctx, query) {
+            Ok(_) | Err(QueryError::PrimaryNotFound(_)) => Ok(Vec::new()),
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn initial_override_value(
+        &self,
+        _ctx: &ViewerContext<'_>,
+        _query: &re_data_store::LatestAtQuery,
+        _store: &re_data_store::DataStore,
+        entity_path: &re_log_types::EntityPath,
+        component: &re_types::ComponentName,
+    ) -> Option<re_log_types::DataCell> {
+        if *component == Color::name() {
+            Some([initial_override_color(entity_path)].into())
+        } else {
+            None
+        }
+    }
+}
+
+impl SeriesLineSystem {
+    fn load_scalars(
+        &mut self,
+        ctx: &ViewerContext<'_>,
+        query: &ViewQuery<'_>,
+    ) -> Result<(), QueryError> {
+        re_tracing::profile_function!();
+
+        let query_caches = ctx.entity_db.query_caches();
+        let store = ctx.entity_db.store();
+
+        let egui_ctx = &ctx.re_ui.egui_ctx;
+
+        let plot_mem = egui_plot::PlotMemory::load(egui_ctx, crate::plot_id(query.space_view_id));
+        let plot_bounds = plot_mem.as_ref().map(|mem| *mem.bounds());
+        // What's the delta in value of X across two adjacent UI points?
+        // I.e. think of GLSL's `dpdx()`.
+        let plot_value_delta = plot_mem.as_ref().map_or(1.0, |mem| {
+            1.0 / mem.transform().dpos_dvalue_x().max(f64::EPSILON)
+        });
```

Comment:
this is the code that determines the aggregation factor and as such I'm not super happy having it triplicated even if the critical part is relatively short. Can we further integrate this between the visualizers?

Reply:
good call